### PR TITLE
Add the `bufferExporter`

### PR DESCRIPTION
**Pull request by MrAlias**
The batching log processor needs to be able to export payloads without blocking the caller. This adds a `bufferExporter` type that will buffer export requests to a chan and perform exports from a spawned goroutine. The `Export`, `ForceFlush`, and `Shutdown` methods are all synchronous calls to the export goroutine (they will not return until their operation is complete). However the added `EnqueueExport` method allows the batching log processor to enqueue exports for asynchronous processing.

Running all exports in their own, single, goroutine ensures that all calls to the Exporter's Export method are called synchronously (required by the interface).

Part of #5063 

See https://github.com/open-telemetry/opentelemetry-go/pull/5093 for the implementation use.

### CODE REVIEW

**Comment by MrAlias** [comment]
Working on tests.

**Review by dmathieu — APPROVED** [review]

File: sdk/log/exporter.go [hunk 1]
```diff
@@ -127,3 +131,132 @@ func (e exportData) respond(err error) {
...
+
+// newBufferExporter returns a new bufferExporter that wraps exporter. The
+// returned bufferExporter will buffer at most size number of export requests.
+// If size is less than zero, zero will be used (i.e. only synchronous
+// exporting will be supported).
+func newBufferExporter(exporter Exporter, size int) *bufferExporter {
+	if size < 0 {
+		size = 0
+	}
+	input := make(chan exportData, size)
```
[the remark this answers is not in the document]
Comment (MrAlias) [reply sdk/log/exporter.go:155]: How do you handle the tests that need different sized buffers?

[the remark this answers is not in the document]
Comment (pellared) [reply sdk/log/exporter.go:155]: I see that zero length buffer is handy for testing. Resolving.

File: sdk/log/exporter.go [hunk 2]
```diff
@@ -127,3 +131,132 @@ func (e exportData) respond(err error) {
...
+	data := exportData{ctx, records, rCh}
+
+	e.inputMu.Lock()
+	defer e.inputMu.Unlock()
+
+	// Check stopped before enqueueing now that e.inputMu is held. This
+	// prevents sends on a closed chan when Shutdown is called concurrently.
+	if e.stopped.Load() {
+		return errStopped
+	}
```
Comment (pellared) [inline sdk/log/exporter.go:176]: We can also add this (second time) at the beginning of the function to reduce the overhead of acquiring a lock when it is not needed (double-checked locking).

**Review by pellared — APPROVED** [review]

File: sdk/log/exporter.go (hunk 2 above)
Comment (MrAlias) [reply sdk/log/exporter.go:176]: I was on the fence about adding that.

Doing so will impact all calls to enqueue, the vast majority of which will not be happening after shutdown. Given those calls are going to be in the hot path, and the most important performance pathway is the OnEmit one, I tried to minimize computation time and not do the extra atomic load.

On the rarer occasions where this called after a shutdown, the performance hit of acquiring a lock prior to an atomic load does not seem like something that will impact users in a meaningful way.

Have I missed something?

File: sdk/log/exporter.go [hunk 3]
```diff
@@ -127,3 +131,132 @@ func (e exportData) respond(err error) {
...
+	e.inputMu.Lock()
+	defer e.inputMu.Unlock()
+
+	// No more sends will be made.
+	close(e.input)
+	select {
+	case <-e.done:
+	case <-ctx.Done():
+		return errors.Join(ctx.Err(), e.Shutdown(ctx))
```
[the remark this answers is not in the document]
Comment (MrAlias) [reply sdk/log/exporter.go]: Fixed, and tests added.

**Pull request merged**